Web Inspector: Font Panel: Avoid needless refresh of FontStyles

rcaliman-apple opened the pull request:
#### 202e031fdd6bde16c53b3817fb7fe229cbcb1884
<pre>
Web Inspector: Font Panel: Avoid needless refresh of FontStyles
<a href="https://bugs.webkit.org/show_bug.cgi?id=250128">https://bugs.webkit.org/show_bug.cgi?id=250128</a>

Reviewed by Patrick Angle.

There are a number of expensive operations in the Fonts sidebar panel done in reaction
to style changes of the inspected node, like computing applicable font properties and their values,
and updating the Fonts panel DOM structure.

These operations must be done whenever styles change from outside the Fonts panel
so it can reflect the latest changes. But they don&apos;t have to be done live.
Throtting the update in reaction to changes from the outside is acceptable.

The slider inputs for editing variation axis values from the Fonts sidebar panel
are a source of high-frequency style changes.
This is necessary so that the inspected page reacts immediately to style changes.
But there&apos;s no need for the Fonts panel to react to style chagnes it originates.
In this scenario it is acceptable to inhibit updates of the Fonts panel and keep state local.

* Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js:
(WI.FontDetailsPanel):
(WI.FontDetailsPanel.prototype.refresh):
(WI.FontDetailsPanel.prototype.update):
(WI.FontDetailsPanel.prototype._handleFontVariationValueChanged):
* Source/WebInspectorUI/UserInterface/Views/FontVariationDetailsSectionRow.js:
(WI.FontVariationDetailsSectionRow):

Canonical link: <a href="https://commits.webkit.org/259821@main">https://commits.webkit.org/259821@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7658a77d6c7da02f70c0933bfc80d682e9fc95ea

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/106072 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/15127 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/38903 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/115253 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/175347 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/16553 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/6296 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/98280 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/114970 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/111825 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/12597 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95577 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/40142 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/94468 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/27220 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/81820 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/8377 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/28572 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/8870 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/5119 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/14490 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/48116 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6796 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/10414 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Views/FontVariationDetailsSectionRow.js:
```diff
@@ -59,6 +59,7 @@ WI.FontVariationDetailsSectionRow = class FontVariationDetailsSectionRow extends
         this._variationMaxValueElement.textContent = this._formatAxisValueAsString(maximumValue);
 
         this._inputRangeElement.addEventListener("input", (event) => {
+            this.value = event.target.value;
```
This keeps the perception to a live Fonts panel UI while a user is dragging on a slider. State is update locally and then synchronized later once the debounced update occurs.

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js:
```diff
@@ -521,6 +536,8 @@ WI.FontDetailsPanel = class FontDetailsPanel extends WI.StyleDetailsPanel
 
     _handleFontVariationValueChanged(event)
     {
+        this._skipNextUpdate = true;
```
The goal is to prevent DOM thrashing in the Fonts sidebar panel while dragging a slider. 
Noticeable by a jittery slider thumb. No need to update it with the value it itself originated. 
While dragging, the value from the model is always slightly behind the current value of the slider.

patrickangle replied:
Won't the throttling for updates already reduce DOM thrashing? I'm not sure its worth missing incoming updates that may not have originated from us to save an operation every quarter of a second (or every 10th/20th of a second if we reduce the intervals)

rcaliman-apple replied:
> Won't the throttling for updates already reduce DOM thrashing?

Not quite. Even with throttling you can get into the scenario of updating the value for the slider you're currently dragging. It just happens less frequently. While dragging, the reaction to change is always for an older value, hence causing jitter. Not the end of the world, but it's there.

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js:
```diff
@@ -521,6 +536,8 @@ WI.FontDetailsPanel = class FontDetailsPanel extends WI.StyleDetailsPanel
     _handleFontVariationValueChanged(event)
     {
+        this._skipNextUpdate = true;
+        this._debouncedUpdate.delayForTime(250);
```
A quarter of a second is an awfully long time between updates IMO when trying to view changes on the page. Long enough I'd think that I'd overshoot where I want to land with the slider if I'm trying to drag to a specific value. Is performance acceptable around 50ms or 100ms?

rcaliman-apple replied:
Updates to the page styles always happen live. They are not debounced or throttled.
What we're handing in this patch is the Fonts panel **reacting to style changes**, some of which it originates.

> Long enough I'd think that I'd overshoot where I want to land with the slider if I'm trying to drag to a specific value.

The values in the Fonts panel, Styles panel and what's visible on page remain in sync throughout slider interaction. There's no chance of overshooting.

The Fonts panel value in the DOM is updated locally by the slider's `input` event handler before dispatching the event to apply the styles to the page. 

`FontVariationDetailsSectionRow.js:61-64`
```js
this._inputRangeElement.addEventListener("input", (event) => {
    this.value = event.target.value;
    this.dispatchEventToListeners(WI.FontVariationDetailsSectionRow.Event.VariationValueChanged, {tag: event.target.name, value: event.target.value});
}, {signal: abortSignal});
 ```
 
Here's what's going on while dragging a slider:

```
this._skipNextUpdate = true
```
Flipping this flag inhibits updating the Fonts panel DOM and doing associated `FontStyles` computations *in reaction to changes* we're fairly certain we just caused.

```
this._debouncedUpdate.delayForTime(250)
```
A short time after the slider stops changing (`250ms` in this case; could be lowered) we force an update just in case there's a style change that occurred in the page after the Fonts panel stopped making changes. This is the "failsafe" we discussed to ensure we catch any updates by in-page scripting.

In the debounced operation, the `this._skipNextUpdate` flag is flipped back off. This allows the Fonts panel to update in reaction to changes coming from the outside. These updates are throttled.

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js:
```diff
@@ -38,6 +38,9 @@ WI.FontDetailsPanel = class FontDetailsPanel extends WI.StyleDetailsPanel
         this._basicPropertyNames = ["font-size", "font-style", "font-weight", "font-stretch"];
 
         this._abortController = new AbortController;
+        this._throttledUpdate = new Throttler(() => { this.update() }, 250);
```
I also think this is too slow, since it will also affect programmatic changes an author might be making. Only updating every quarter of a second will feel like a performance issue. I'd again consider 50-100ms max here.

rcaliman-apple replied:
You can argue that the Fonts panel should react more frequently to style changes caused outside of it to give the illusion of performance, while, ironically, degrading it a bit :) 

I'm skeptical that reflecting instant updates is the intent of someone monitoring the Fonts panel while the page styles are actively changing, user-driven or scripted.

When making high-frequency changes in the Styles panel (increment/decrement using the arrow keys), my hypothesis is that a developer is monitoring the page itself to observe the effects and the value in the Styles panel to know when to stop. A slight delay to another panel, the Fonts panel, is likely not their focus. Even so, throttling ensures the panel updates at reasonably quick intervals.

If you prefer, we can lower the throttling rate to `100ms`.

I'm not too vested in this throttling as much as I am with preventing reacting to style changes that originate from the Fonts panel, the `this._skipNextUpdate` flag. Range inputs are a source of high-frequency changes we want to control.

patrickangle reviewed (commented):
Apologies it took me a few days to get to reviewing this, but I have some concerns about how big the intervals are here, as well as the potential to miss incoming updates that did not originate from the frontend (e.g. from the pages JS).

patrickangle reviewed (approved):
Thank you for clarifying and helping me understand some of the subtle details. lgtm

rcaliman-apple committed:
Web Inspector: Font Panel: Avoid needless refresh of FontStyles
https://bugs.webkit.org/show_bug.cgi?id=250128

Reviewed by Patrick Angle.

There are a number of expensive operations in the Fonts sidebar panel done in reaction
to style changes of the inspected node, like computing applicable font properties and their values,
and updating the Fonts panel DOM structure.

These operations must be done whenever styles change from outside the Fonts panel
so it can reflect the latest changes. But they don't have to be done live.
Throtting the update in reaction to changes from the outside is acceptable.

The slider inputs for editing variation axis values from the Fonts sidebar panel
are a source of high-frequency style changes.
This is necessary so that the inspected page reacts immediately to style changes.
But there's no need for the Fonts panel to react to style chagnes it originates.
In this scenario it is acceptable to inhibit updates of the Fonts panel and keep state local.

* Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js:
(WI.FontDetailsPanel):
(WI.FontDetailsPanel.prototype.refresh):
(WI.FontDetailsPanel.prototype.update):
(WI.FontDetailsPanel.prototype._handleFontVariationValueChanged):
* Source/WebInspectorUI/UserInterface/Views/FontVariationDetailsSectionRow.js:
(WI.FontVariationDetailsSectionRow):

Canonical link: https://commits.webkit.org/259821@main
